### PR TITLE
Add patched server interactions

### PR DESCRIPTION
**Pull request by CTY-git**

### CODE REVIEW

**Comment by patched-admin** [comment]
This patch introduces a new command line option and logic in app.py for connecting to the Patched API, integrates the PatchedClient class for telemetry, and adds custom classes and methods to patched.py such as a connection pool, client class, and functions for patchflow runs. It also includes error logging and handling for connection failures. Additionally, a new function `get_current_branch` is added to utils.py to determine the current branch, replacing the existing git command method in CommitChanges.py.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [This patch in the app.py file adds a new command line option "--patched_api_key" and logic to handle the input value for connecting to Patched API. Additionally, it integrates the PatchedClient class to the application to use the Patched API key for telemetry.](https://github.com/patched-codes/patchwork/pull/131/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/patched.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-0998e5a615567a2c44e0de84a938c4d225e27ae97018e06f32775aa08db9a091)
  [This patch adds new classes and methods to the patched.py file, including a custom HTTPS connection pool with TCP keep-alive settings, a client class for interacting with a specific service, and functions for testing and recording patchflow runs. The changes also include error logging and handling for connection failures and exceptions.](https://github.com/patched-codes/patchwork/pull/131/files#diff-0998e5a615567a2c44e0de84a938c4d225e27ae97018e06f32775aa08db9a091)

</div>

<div markdown="1">

* File changed: [patchwork/common/utils.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-64d094a0b6e480bc072eadafd1516f4c70f777242a3b68df96ec8e263174e35d)
  [The diff adds a new function `get_current_branch` to utils.py which takes a `Repo` object as an argument and returns the current branch. The function checks for a detached head state, retrieves the branch information, and raises an error if the branch cannot be determined.](https://github.com/patched-codes/patchwork/pull/131/files#diff-64d094a0b6e480bc072eadafd1516f4c70f777242a3b68df96ec8e263174e35d)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CommitChanges/CommitChanges.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-96ac14efb9a0f18811385c844af6d763ea1230c095667901bfffdfcf06af9fe6)
  [In this patch to the `CommitChanges.py` file, a new function `get_current_branch` is imported from `patchwork.common.utils`, and the existing `get_current_branch` function that determines the current branch using git commands is removed.](https://github.com/patched-codes/patchwork/pull/131/files#diff-96ac14efb9a0f18811385c844af6d763ea1230c095667901bfffdfcf06af9fe6)

</div>

File: patchwork/app.py [hunk 1]
```diff
@@ -112,15 +114,27 @@ def list_option_callback(ctx: click.Context, param: click.Parameter, value: str
 @click.option(
     "data_format", "--format", type=click.Choice(["yaml", "json"]), default="json", help="Format of the output file."
 )
-def cli(log: str, patchflow: str, opts: list[str], config: str | None, output: str | None, data_format: str):
+@click.option("patched_api_key", "--patched_api_key", help="API key for to connect with patched.")
```
Comment (codelion) [inline patchwork/app.py:117]: "API key to use with the patched.codes service."

**Review by codelion — APPROVED** [review]

**Comment by patched-admin** [comment]
The app.py file in the patchwork application has been updated to include a new option "--patched_api_key" in the CLI interface for users to provide an API key for the patched.codes service. There are also new classes and functions introduced for handling HTTP connections, HTTPS connections with TCP keep-alive functionality, and a customized PoolManager for a PatchedClient. Additionally, a new function `get_current_branch` has been added to the `utils.py` file in the `common` directory to handle different cases for detached and non-detached head states when getting the current branch of a repository, and the function for getting the current branch in the CommitChanges.py file has been removed in favor of importing it from the patchwork.common.utils module using the get_current_branch function.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [The code in the app.py file of the patchwork application has been updated to include a new option "--patched_api_key" in the CLI interface, allowing users to provide an API key for the patched.codes service. This API key is then utilized in the code to initialize a PatchedClient instance with the provided key for telemetry purposes.](https://github.com/patched-codes/patchwork/pull/131/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/patched.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-0998e5a615567a2c44e0de84a938c4d225e27ae97018e06f32775aa08db9a091)
  [The diff introduces new classes and functions to handle HTTP connections in a Python file. It includes classes for handling HTTPS connections with TCP keep-alive functionality, a customized PoolManager, and ParamType for a PatchedClient that interacts with a Patched web application including token validation and recording of patchflow runs.](https://github.com/patched-codes/patchwork/pull/131/files#diff-0998e5a615567a2c44e0de84a938c4d225e27ae97018e06f32775aa08db9a091)

</div>

<div markdown="1">

* File changed: [patchwork/common/utils.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-64d094a0b6e480bc072eadafd1516f4c70f777242a3b68df96ec8e263174e35d)
  [The diff adds a new function `get_current_branch` to the `utils.py` file in the `common` directory. This function takes a `Repo` object as input and returns the current branch of the repository, handling different cases for detached and non-detached head states.](https://github.com/patched-codes/patchwork/pull/131/files#diff-64d094a0b6e480bc072eadafd1516f4c70f777242a3b68df96ec8e263174e35d)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CommitChanges/CommitChanges.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-96ac14efb9a0f18811385c844af6d763ea1230c095667901bfffdfcf06af9fe6)
  [The diff in the CommitChanges.py file removes the function for getting the current branch directly from a Repo object and instead imports it from the patchwork.common.utils module using the get_current_branch function.](https://github.com/patched-codes/patchwork/pull/131/files#diff-96ac14efb9a0f18811385c844af6d763ea1230c095667901bfffdfcf06af9fe6)

</div>

**Comment by patched-admin** [comment]
The code changes in app.py add a new import statement for PatchedClient and Repo, along with a new command line option `patched_api_key`, which is used if provided in PatchedClient initialization, and in a PatchedClient method within the cli function's try block. In patched.py, classes related to managing HTTP connections are introduced, including TCPKeepAliveHTTPSConnectionPool and KeepAlivePoolManager, used in the PatchedClient class for handling HTTP requests and managing patchflow runs. The `utils.py` change adds a `get_current_branch` function using the `git` library to retrieve the current branch from a Git repository with proper error handling for detached HEAD state scenarios. In CommitChanges.py, the get_current_branch function is imported from utils module, the implementation is removed, and `repo` initialization in the run method now includes search_parent_directories=True for the git.Repo constructor. Finally, in CreatePR.py, the code is updated to use the `search_parent_directories` parameter when initializing `git.Repo`, enabling it to search for repositories in parent directories if necessary.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [The code in app.py adds a new import statement for PatchedClient and Repo, includes a new command line option called `patched_api_key`, uses the `patched_api_key` if provided in the PatchedClient initialization, and uses the Repo object along with the `patched_api_key` in a PatchedClient method within the cli function's try block to handle exceptions appropriately.](https://github.com/patched-codes/patchwork/pull/131/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/patched.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-0998e5a615567a2c44e0de84a938c4d225e27ae97018e06f32775aa08db9a091)
  [This diff introduces classes related to managing HTTP connections in the patched.py file, including TCPKeepAliveHTTPSConnectionPool and KeepAlivePoolManager. It also adds a PatchedClient class with methods for handling HTTP requests and managing patchflow runs, utilizing the added classes for connection management.](https://github.com/patched-codes/patchwork/pull/131/files#diff-0998e5a615567a2c44e0de84a938c4d225e27ae97018e06f32775aa08db9a091)

</div>

<div markdown="1">

* File changed: [patchwork/common/utils.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-64d094a0b6e480bc072eadafd1516f4c70f777242a3b68df96ec8e263174e35d)
  [This diff in the `utils.py` file adds a new function `get_current_branch` which retrieves the current branch from a Git repository using the `git` library. It addresses cases where the repository may be in a detached HEAD state with additional commits by providing appropriate error handling.](https://github.com/patched-codes/patchwork/pull/131/files#diff-64d094a0b6e480bc072eadafd1516f4c70f777242a3b68df96ec8e263174e35d)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CommitChanges/CommitChanges.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-96ac14efb9a0f18811385c844af6d763ea1230c095667901bfffdfcf06af9fe6)
  [This diff in CommitChanges.py introduces an import statement for the get_current_branch function from patchwork.common.utils module, and removes the implementation of the get_current_branch function from the file. Additionally, it changes the initialization of the 'repo' variable in the run method to use the search_parent_directories=True option for the git.Repo constructor.](https://github.com/patched-codes/patchwork/pull/131/files#diff-96ac14efb9a0f18811385c844af6d763ea1230c095667901bfffdfcf06af9fe6)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CreatePR/CreatePR.py](https://github.com/patched-codes/patchwork/pull/131/files#diff-6f7d9cad8d3bb8c511281e360b5038717dc6590c402e597ab408aec045a6d5d4)
  [The code in the `CreatePR.py` file has been modified to use the `search_parent_directories` parameter when initializing the `git.Repo` object, allowing it to search for the repository in parent directories if needed.](https://github.com/patched-codes/patchwork/pull/131/files#diff-6f7d9cad8d3bb8c511281e360b5038717dc6590c402e597ab408aec045a6d5d4)

</div>

**Pull request merged**